doc/design/programming_solver.md

```diff
@@ -0,0 +1,141 @@
+
+- [Linear programming](https://en.wikipedia.org/wiki/Linear_programming)
+- [Quadratic programming](https://en.wikipedia.org/wiki/Quadratic_programming)
+- [Nonlinear programming](https://en.wikipedia.org/wiki/Nonlinear_programming)
+- [Integer programming](https://en.wikipedia.org/wiki/Integer_programming)
+- [Mixed-integer linear programming](http://macc.mcmaster.ca/maccfiles/chachuatnotes/07-MILP-I_handout.pdf)
+- [Mixed-integer quadratic programming](http://www.optimization-online.org/DB_FILE/2014/07/4446.pdf)
+- [Mixed-integer nonlinear programming](https://link.springer.com/book/10.1007/978-1-4614-1927-3)
+
+## Background
+
+To understand what's a mathematical programming problem, let's take a look at this example:
+
+Giapetto’s Woodcarving, Inc., manufactures two types of wooden toys: soldiers and trains. 
+
+A soldier sells for $27 and uses $10 worth of raw materials.  Each soldier that is manufactured increases Giapetto’s variable labor and overhead costs by $14.  A train sells for $21 and uses $9 worth of raw materials.  Each train built increases Giapetto’s variable labor and overhead costs by $10.  The manufacture of wooden soldiers and trains requires two types of skilled labor: carpentry and finishing.  A soldier requires 2 hours of finishing labor and 1 hour of carpentry labor.  A train requires 1 hour of finishing labor and 1 hour of carpentry labor.  Each week, Giapetto can obtain all the needed raw material but only 100 finishing hours and 80 carpentry hours.  Demand for trains is unlimited, but at most 40 soldiers are bought each week.  Giapetto wants to maximize weekly profit (revenues-costs).
+
+Let
+
+- x be the number of soldiers produced each week
+- y be the number of trains produced each week
+
+Then the objective is: 
+
+**Maximise Z = (27 - 10 - 14)x + (21 - 9 - 10)y = 3x + 2y**
+
+The constraints are:
+
+- 2x + y <= 100 (Finishing Constraint)
+- x + y <= 80 (Carpentry Constraint)
+- x <=40 (Constraint on demand for soldiers)
+- x,y >= 0 (sign restriction)
+
+
+To solve this problem, we can write objective and constraints using SQL, then SQLFlow should call the corresponding "solver" (e.g. [pyomo](http://www.pyomo.org/) and [GLPK](https://www.gnu.org/software/glpk/)) to find the result and output the result into the result table.
+
+## Grammar Design
+
+Generally, SQLFlow users can write one SQL statement to describe the programming problem:
+
+```sql
+SELECT * FROM train_table
+TO SOLVE LP
+WITH
+     objective="sum([(train_table.price[i] - train_table.materials_cost[i] - train_table.other_cost[i]) * @X[i] for i in @X])",
+     constraints=["sum([train_table.finishing[i] * @X[i] for i in @X]) <= 100",
+                  "sum([train_table.carpentry[i] * @X[i] for i in @X]) <= 80",
+                  "@X[i] <= train_table.max_num[i] for i in @X",
+                  "@X[i] >= 0 for i in @X"],
```

Comment:
Why we do require users to write constraints and the objective as strings with the above syntax? Is that because pyomo accepts this syntax?  If that is the reason, how if SQLFlow is going to connect to other mathematical programming systems, which accept different problem description syntax?

Reply:
> Why we do require users to write constraints and the objective as strings with the above syntax? Is that because pyomo accepts this syntax?

Yes, `pyomo` defines an AML (Algebraic Modeling Language) which is popular and simple to use, as described here: https://pyomo.readthedocs.io/en/stable/pyomo_overview/math_modeling.html#programming-language

> Pyomo leverages a high-level programming language, which has several advantages over custom AMLs: a very robust language, extensive documentation, a rich set of standard libraries, support for modern programming features like classes and functions, and portability to many platforms.

If we are going to connect to other mathematical programming systems, we can write a different code generator like what we already did when connecting to Tensorflow and XGboost.

Reply:
It seems that if other mathematical programming systems can accept AML rules, we can allow SQLFlow users to encode AML programs as string literals in SQLFlow programs like described in this example.  Otherwise, we might prefer to extend SQLFlow syntax to do what AML describes, and make SQLFlow compiler/codegen generate AML rules.

Should we talk with authors of other mathematical programming tools?